### PR TITLE
Add Yeahmobi bidder and tests

### PR DESCRIPTION
**Pull request by AndriyPavlyuk**
Adding new implementation of Yeahmobi bidder to keep up with Go version.

### CODE REVIEW

File: src/test/resources/org/prebid/server/it/openrtb2/yeahmobi/test-auction-yeahmobi-response.json [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+              "cache": {
+                "bids": {
+                  "cacheId": "f0ab9105-cb21-4e59-b433-70f5ad6671cb",
+                  "url": "{{ cache.resource_url }}f0ab9105-cb21-4e59-b433-70f5ad6671cb"
+                }
+              }
+            }
+          },
+          "id": "1",
+          "impid": "imp123",
```
Comment (BraslavskiyAndrey) [inline src/test/resources/org/prebid/server/it/openrtb2/yeahmobi/test-auction-yeahmobi-response.json:37]: In request you have impId = impId001, in response imp123. The same for cache and bid response json files

Comment (AndriyPavlyuk) [reply src/test/resources/org/prebid/server/it/openrtb2/yeahmobi/test-auction-yeahmobi-response.json:37]: Yeah, thank you for notice and attention!
This is the mistake in test-yeahmobi-bid-response.json. That caused bad results in other test json's.
I fixed it.

File: src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            updateNativeRequest = nativeRequest != null && nativeRequest.get("native") == null
+                    ? mapper.mapper().writeValueAsString(objectNode)
+                    : null;
+        } catch (JsonProcessingException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+
+        return updateNativeRequest != null
+                ? imp.toBuilder().xNative(Native.builder().request(updateNativeRequest).build()).build()
+                : imp;
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java:97]: They have different logic in GO. 
We need to work with json that in `native.request`.
```
  "native": {
          "request": "{\"native\":{}}\"
}
```
So if there are no imp.native we do not need to do anything. 
If there are no imp.native.request JSON NATIVE we need to add this object. You can check their tests to fix our IT tests.

File: src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+    private HttpRequest<BidRequest> createSingleRequest(Imp imp, BidRequest request, String url) {
+        final BidRequest outgoingRequest = request.toBuilder().imp(Collections.singletonList(imp)).build();
+
+        final String body = mapper.encode(outgoingRequest);
+
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(url)
+
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java:108]: redundant

File: src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        final List<HttpRequest<BidRequest>> result = new ArrayList<>();
+
+        for (Imp imp : request.getImp()) {
+            try {
+                final ExtImpYeahmobi extImpYeahmobi = parseImpExt(imp);
+                final Imp processImp = processImp(imp);
+                final String url = endpointUrl + "gw-" + extImpYeahmobi.getZoneId() + "-bid.yeahtargeter.com";
+                result.add(createSingleRequest(processImp, request, url));
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java:62]: They do not creating several request. They only creating one

File: src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            throw new PreBidException(e.getMessage(), e);
+        }
+
+        return updateNativeRequest != null
+                ? imp.toBuilder().xNative(Native.builder().request(updateNativeRequest).build()).build()
+                : imp;
+    }
+
+    private HttpRequest<BidRequest> createSingleRequest(Imp imp, BidRequest request, String url) {
+        final BidRequest outgoingRequest = request.toBuilder().imp(Collections.singletonList(imp)).build();
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java:101]: bug

File: src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        final List<HttpRequest<BidRequest>> result = new ArrayList<>();
+
+        for (Imp imp : request.getImp()) {
+            try {
+                final ExtImpYeahmobi extImpYeahmobi = parseImpExt(imp);
+                final Imp processImp = processImp(imp);
+                final String url = endpointUrl + "gw-" + extImpYeahmobi.getZoneId() + "-bid.yeahtargeter.com";
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java:61]: In production you'll get ` https://{{Host}}/prebid/bidgw-zoneId-bid.yeahtargeter.com`

You forget to replace {{Host}}, not just add values.

File: src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            if (imp.getId().equals(impId)) {
+                if (imp.getBanner() != null) {
+                    break;
+                } else if (imp.getVideo() != null) {
+                    bidType = BidType.video;
+                    break;
+                } else if (imp.getXNative() != null) {
+                    bidType = BidType.xNative;
+                    break;
+                }
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java:164]: just  use `return` with desirable value

File: src/test/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidderTest.java [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        final Result<List<HttpRequest<BidRequest>>> result = yeahmobiBidder.makeHttpRequests(bidRequest);
+
+        // then
+        assertThat(result.getErrors()).hasSize(1);
+        assertThat(result.getErrors().get(0).getMessage()).startsWith("Cannot deserialize instance");
+        assertThat(result.getValue()).isEmpty();
+    }
+
+    @Test
+    public void makeHttpRequestsShouldCreateCorrectURL() {
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidderTest.java:70]: changed bc of replacement of host

File: src/test/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidderTest.java [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+        final Result<List<HttpRequest<BidRequest>>> result = yeahmobiBidder.makeHttpRequests(bidRequest);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).hasSize(1);
+        assertThat(result.getValue().get(0).getUri()).isEqualTo("https://test.endpoint.comgw-zoneId-bid.yeahtargeter.com");
+    }
+
+    @Test
+    public void makeHttpRequestsShouldNotChangeBannerWidthAndHeightIfPresent() {
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidderTest.java:88]: name != what you are testing

File: src/test/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidderTest.java [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+        // given
+        String nativeRequest = "{\"ver\":\"1.2\",\"context\":1,\"plcmttype\":4,\"plcmtcnt\":1,\"assets\":[{\"id\":2,"
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidderTest.java:90]: check Go tests, they have different value in there

File: src/test/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidderTest.java [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+                + ":128,\"hmin\":128,\"mimes\":[\"image/jpg\",\"image/jpeg\",\"image/png\"]}},{\"id\":7,"
+                + "\"required\":1,\"data\":{\"type\":2,\"len\":120}}]}";
+
+        final BidRequest bidRequest = givenBidRequest(
+                impBuilder -> impBuilder
+                        .banner(Banner.builder()
+                                .format(singletonList(Format.builder().w(300).h(500).build()))
+                                .w(200)
+                                .h(150)
+                                .build())
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidderTest.java:101]: redundant

File: src/test/resources/org/prebid/server/it/openrtb2/yeahmobi/test-yeahmobi-bid-request.json [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+      "banner": {
+        "format": [
+          {
+            "w": 300,
+            "h": 250
+          }
+        ]
+      },
+      "native": {
+        "request": "{\"native\":{\"ver\":\"1.2\",\"context\":1,\"plcmttype\":4,\"plcmtcnt\":1,\"assets\":[{\"id\":2,\"required\":1,\"title\":{\"len\":90}},{\"id\":6,\"required\":1,\"img\":{\"type\":3,\"wmin\":128,\"hmin\":128,\"mimes\":[\"image/jpg\",\"image/jpeg\",\"image/png\"]}},{\"id\":7,\"required\":1,\"data\":{\"type\":2,\"len\":120}}]}}"
```
Comment (DGarbar) [inline src/test/resources/org/prebid/server/it/openrtb2/yeahmobi/test-yeahmobi-bid-request.json:15]: You need to get smth like that 
`{"native":{"assets":[{"id":2,"required":1,"title":{"len":90}},{"id":6,"img":{"hmin":128,"mimes":["im`

Comment (AndriyPavlyuk) [reply src/test/resources/org/prebid/server/it/openrtb2/yeahmobi/test-yeahmobi-bid-request.json:15]: In Go they do not have any sorting operations and it doesn't matter order

File: src/test/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidderTest.java (hunk 10 above)
Comment (DGarbar) [reply src/test/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidderTest.java:90]: Ok. This example is working for both GO and current implementation.

File: src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java [hunk 13]
```diff
@@ -52,20 +52,38 @@ public YeahmobiBidder(String endpointUrl, JacksonMapper mapper) {
...
+                    extImpYeahmobi = parseImpExt(imp);
+                }
                 final Imp processImp = processImp(imp);
-                final String url = endpointUrl + "gw-" + extImpYeahmobi.getZoneId() + "-bid.yeahtargeter.com";
-                result.add(createSingleRequest(processImp, request, url));
+                validImps.add(processImp);
             } catch (PreBidException e) {
                 errors.add(BidderError.badInput(e.getMessage()));
             }
         }
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java:69]: You need to return error when there are no ExtImpYeahmobi

File: src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java [hunk 14]
```diff
@@ -52,20 +52,38 @@ public YeahmobiBidder(String endpointUrl, JacksonMapper mapper) {
...
 
-        return Result.of(result, errors);
+        final String host = extImpYeahmobi != null ? String.format("gw-%s-bid.yeahtargeter.com",
+                HttpUtil.encodeUrl(extImpYeahmobi.getZoneId())) : null;
+        final String url = host != null ? endpointUrl.replace("{{Host}}", host) : null;
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java:72]: Will be simplified after fix above

File: src/test/java/org/prebid/server/it/YeahmobiTest.java [hunk 15]
```diff
@@ -25,7 +25,7 @@
     public void openrtb2AuctionShouldRespondWithBidsFromYeahmobi() throws IOException, JSONException {
         // given
         // AdmixerBidder bid response for imp 001
-        WIRE_MOCK_RULE.stubFor(post(urlPathEqualTo("/yeahmobi-exchange/gw-sin-bid.yeahtargeter.com"))
+        WIRE_MOCK_RULE.stubFor(post(urlPathEqualTo("/yeahmobi-exchange/"))
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/it/YeahmobiTest.java:28]: `gw-sin-bid.yeahtargeter.com"` will be added after above fix

File: src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java [hunk 16]
```diff
@@ -57,19 +57,21 @@ public YeahmobiBidder(String endpointUrl, JacksonMapper mapper) {
...
                 errors.add(BidderError.badInput(e.getMessage()));
             }
         }
 
-        final String host = extImpYeahmobi != null ? String.format("gw-%s-bid.yeahtargeter.com",
-                HttpUtil.encodeUrl(extImpYeahmobi.getZoneId())) : null;
-        final String url = host != null ? endpointUrl.replace("{{Host}}", host) : null;
+        if (extImpYeahmobi == null) {
+            errors.add(BidderError.badInput("Invalid ExtImpYeahmobi value"));
+            return Result.of(Collections.emptyList(), errors);
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/yeahmobi/YeahmobiBidder.java:70]: Use special method. With errors

File: src/test/java/org/prebid/server/it/YeahmobiTest.java [hunk 17]
```diff
@@ -24,8 +24,8 @@
     @Test
     public void openrtb2AuctionShouldRespondWithBidsFromYeahmobi() throws IOException, JSONException {
         // given
-        // AdmixerBidder bid response for imp 001
-        WIRE_MOCK_RULE.stubFor(post(urlPathEqualTo("/yeahmobi-exchange/"))
+        // YeahmobiBidder bid response for imp 001
+        WIRE_MOCK_RULE.stubFor(post(urlPathEqualTo("/yeahmobi-exchange"))
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/it/YeahmobiTest.java:28]: Are you sure that /gw-%s-bid.yeahtargeter.com not there ?

Comment (AndriyPavlyuk) [reply src/test/java/org/prebid/server/it/YeahmobiTest.java:28]: Not there bc it doesn't override http://localhost:8090/yeahmobi-exchange

**Pull request merged**